ssl/ssl_lib.c

```diff
@@ -6901,3 +6909,173 @@ int SSL_CTX_set0_tmp_dh_pkey(SSL_CTX *ctx, EVP_PKEY *dhpkey)
     ctx->cert->dh_tmp = dhpkey;
     return 1;
 }
+
+#ifndef OPENSSL_NO_COMP
+
+static int ssl_set_cert_comp_pref(uint16_t *array, size_t array_size, uint16_t *algs, int num_algs)
+{
+    int i, j = 0;
+
+    /* This will 0-terminate the array */
+    memset(array, 0, array_size);
+    /* include only those algorithms we support, ignore the rest if the list fills */
+    for (i = 0; i < num_algs && j < TLSEXT_comp_cert_limit; i++) {
```

Comment:
No, the array has `uint16_t` elements, so that would go twice as far as allowed.

Reply:
(But yes, `j` should be limited by the number of array elements)

Reply:
I changed this around to remove the need for checking `j` by checking for duplicates. That way, only three elements of the 4 element `array` can be filled (`array` is initially `memset` to `0` so that it is also `0` terminated). In addition, a temporary `tmp_prefs` array is used for the calculation, such that the `SSL_CONNECTION`/`SSL_CTX` `cert_comp_prefs` is not modified unless success.